File: 475_Heaters.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/heaters/
# Winter is coming! Your first job during the contest is to design a standard heater with fixed warm radius to warm
# all the houses.
# Now, you are given positions of houses and heaters on a horizontal line, find out minimum radius of heaters so that
# all houses could be covered by those heaters.
# So, your input will be the positions of houses and heaters seperately, and your expected output will be the minimum
# radius standard of heaters.

# Sort heaters and houses. Add sentinel heaters at + and - infinity. For each house, move along the heaters until the
# current heater is on the left of the house and the next heater is at or on the right of the house. One of these
# heaters is the closest to the house. Update the radius to be the max of the current radius and the closest heater.
# Time - O(m log m + n log n)
# Space - O(m) where m = len(heaters)

class Solution(object):
    def findRadius(self, houses, heaters):
        """
        :type houses: List[int]
        :type heaters: List[int]
        :rtype: int
        """
        heaters.sort()
        houses.sort()
        heaters = [float("-inf")] + heaters + [float("inf")]

        i = 0
        radius = -1

        for house in houses:

            while heaters[i + 1] < house:
                i += 1
            left_distance = house - heaters[i]
            right_distance = heaters[i + 1] - house         # could be zero
            closest = min(left_distance, right_distance)
            radius = max(radius, closest)

        return radius

File: java/475_Heaters.java
/*
https://leetcode.com/problems/heaters/
Winter is coming! Your first job during the contest is to design a standard heater with fixed warm radius to warm
all the houses. You are given positions of houses and heaters on a horizontal line, find out minimum radius of heaters
so that all houses could be covered by those heaters.
Your input will be the positions of houses and heaters seperately, and your expected output will be the minimum radius
standard of heaters.
As long as a house is in the heaters' warm radius range, it can be warmed.
All the heaters follow your radius standard and the warm radius will the same.

Sort the houses and heaters in ascending order. For each house, if the next heater is the same distance or closer,
increment the current heater. Then the best closest heater is found, update the minRadius.
Time - O(mlogm + nlogn)
Space - O(m + n)
*/

public class Solution {
    public int findRadius(int[] houses, int[] heaters) {
        Arrays.sort(houses);
        Arrays.sort(heaters);

        int minRadius = 0;
        int j = 0;              // heater index
        int distance = 0;       // between current house and current heater

        for (int house : houses) {
            distance = Math.abs(house - heaters[j]);
            while (j < heaters.length - 1 && Math.abs(house - heaters[j + 1]) <= distance) {
                distance = Math.abs(house - heaters[j + 1]);
                ++j;
            }
            minRadius = Math.max(minRadius, distance);
        }

        return minRadius;
    }
}
